bee-ledger/src/types/balance.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::types::{error::Error, BalanceDiff};
+
+use bee_common::packable::{Packable, Read, Write};
+
+/// Holds the balance of an address.
+#[derive(Debug, Default)]
+pub struct Balance {
+    amount: u64,
+    dust_allowance: u64,
+    dust_outputs: u64,
+}
+
+impl Balance {
+    /// Creates a new `Balance`.
+    pub fn new(amount: u64, dust_allowance: u64, dust_outputs: u64) -> Self {
+        Self {
+            amount,
+            dust_allowance,
+            dust_outputs,
+        }
+    }
+
+    /// Returns the amount of the `Balance`.
+    pub fn amount(&self) -> u64 {
+        self.amount
+    }
+
+    /// Returns the dust allowance of the `Balance`.
+    pub fn dust_allowance(&self) -> u64 {
+        self.dust_allowance
+    }
+
+    /// Returns the number of dust outputs of the `Balance`.
+    pub fn dust_outputs(&self) -> u64 {
+        self.dust_outputs
+    }
+
+    /// Safely applies a `BalanceDiff` to the `Balance`.
+    pub fn apply_diff(self, diff: &BalanceDiff) -> Result<Self, Error> {
+        let amount = (self.amount as i64)
+            .checked_add(diff.amount())
+            .ok_or_else(|| Error::BalanceOverflow(self.amount as i128 + diff.amount() as i128))?;
+        let dust_allowance = (self.dust_allowance() as i64)
+            .checked_add(diff.dust_allowance())
+            .ok_or_else(|| Error::BalanceOverflow(self.dust_allowance() as i128 + diff.dust_allowance() as i128))?;
+        let dust_outputs = (self.dust_outputs as i64)
+            .checked_add(diff.dust_outputs())
+            .ok_or_else(|| Error::BalanceOverflow(self.dust_outputs as i128 + diff.dust_outputs() as i128))?;
```

Comment:
`self.amount/dust_allowance/dust_outputs as i64` will wrap if those values contain values that `> u64::MAX/2`. Are we sure that isn't possible. If not, then better use `as i128` conversions.

Reply:
Yeah it's all already syntactically bound by the number of IOTA tokens but I guess it wouldn't hurt to use i128, will do

Reply:
Actually it's not possible, the checked operations wouldn't work anymore

Reply:
Also, the `self.amount/dust_allowance/dust_outputs` have to be the result of a previous call to `apply_diff ` and we start with all `0` so this just can't happen.